hr_course/models/hr_course.py

```diff
@@ -180,6 +183,53 @@ def cancel_course(self):
         for record in self:
             record.write(record._cancel_course_values())
 
+    @api.model
+    def send_course_notification_email(self):
+        company_id = self.env.context.get("company_id") or self.env.company.id
+        channel_id = (
+            self.env["res.company"].browse(company_id).course_expiration_channel_id
+        )
+        email_list = channel_id.channel_last_seen_partner_ids.mapped("partner_email")
+        company = self.env["res.company"].browse(
+            self.env.context.get("company_id") or self.env.company.id
+        )
+        if email_list:
+            email_template = self.env.ref("hr_course.mail_template_validity_reminder")
+            for email in email_list:
+                options = {
+                    "email_to": email.strip(),
+                    "email_from": company.email.strip(),
+                }
+                self.message_post_with_template(
+                    self.id,
+                    template_id=email_template,
+                    email_values=options,
+                    force_send=True,
+                )
+                return True
```

Comment:
```suggestion
        channel = (
            self.env["res.company"].browse(company_id).course_expiration_channel_id
        )
        email_template = self.env.ref("hr_course.mail_template_validity_reminder")
        
        if channel:
            channel.message_post_with_template(
                    template_id=email_template,
                )
                return True
```
we can just use channel features to post emal
args to be checked

Reply:
fine but channel has users assigned and afais msg is sent to these users - shouldn't we then update channel with users when they are assigned to the course?

Reply:
the message should sent to subscribers of the channel, AFAIN no any additional actions should be done

Reply:
ok, for the record, after clarifying scenario, there will be hr user group assigned manually to channel - no need to worry about assigning course attendances.

Reply:
looking at the mail template content, we should  change `<field name="model_id" ref="model_hr_course" />` to use model_mail_channel instead. So the channel need to have the data about the course (course name, course attendants)... is it correct?

Reply:
having a template for a channel that will map courses than attendees is a little bit complex, and also we will need to send courses through context and extract them in email
try to add res_id and model to params
https://github.com/odoo/odoo/blob/1b248ebd58e0a57b070ebf7d67c6da78351a720a/addons/mail/models/mail_thread.py#L1958
and check if the correct thread is used
https://github.com/odoo/odoo/blob/1b248ebd58e0a57b070ebf7d67c6da78351a720a/addons/mail/wizard/mail_compose_message.py#L256

Reply:
updated but not sure if in good place.

Reply:
@i-vyshnevska  having some problem with template,
```
 Failed to load template "<div>\n                    <p>Hi,</p>\n                    <p>The validity time of the course ${object.name} will end up soon. You maybe need to plan a new one. </p>\n                    <p>The concerned employees are:</p>\n                    % for employee in ${ctx.get('course_attendee_ids').mapped('employee_id.name')}:\n                    <p>${employee}</p>\n                    % endfor\n                </div>\n            " 
Traceback (most recent call last):
  File "/odoo/src/addons/mail/models/mail_render_mixin.py", line 333, in _render_template_jinja
    template = jinja_env.from_string(tools.ustr(template_txt))
  File "/usr/local/lib/python3.7/dist-packages/jinja2/environment.py", line 880, in from_string
    return cls.from_code(self, self.compile(source), globals, None)
  File "/usr/local/lib/python3.7/dist-packages/jinja2/environment.py", line 591, in compile
    self.handle_exception(exc_info, source_hint=source_hint)
  File "/usr/local/lib/python3.7/dist-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.7/dist-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "<unknown>", line 5, in template
  File "/usr/local/lib/python3.7/dist-packages/jinja2/environment.py", line 497, in _parse
    return Parser(self, source, name, encode_filename(filename)).parse()
  File "/usr/local/lib/python3.7/dist-packages/jinja2/parser.py", line 901, in parse
    result = nodes.Template(self.subparse(), lineno=1)
  File "/usr/local/lib/python3.7/dist-packages/jinja2/parser.py", line 883, in subparse
    rv = self.parse_statement()
  File "/usr/local/lib/python3.7/dist-packages/jinja2/parser.py", line 130, in parse_statement
    return getattr(self, 'parse_' + self.stream.current.value)()
  File "/usr/local/lib/python3.7/dist-packages/jinja2/parser.py", line 192, in parse_for
    self.stream.expect('name:in')
  File "/usr/local/lib/python3.7/dist-packages/jinja2/lexer.py", line 388, in expect
    next(self)
  File "/usr/local/lib/python3.7/dist-packages/jinja2/lexer.py", line 359, in __next__
    self.current = next(self._iter)
  File "/usr/local/lib/python3.7/dist-packages/jinja2/lexer.py", line 562, in wrap
    for lineno, token, value in stream:
  File "/usr/local/lib/python3.7/dist-packages/jinja2/lexer.py", line 739, in tokeniter
    name, filename)
jinja2.exceptions.TemplateSyntaxError: unexpected char '$' at 258
```
can't spot what's wrong there. Could you have a look? Probably some silly typo.